rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+
+This implies that the public-facing registry code will live in `gravitational/next`.
+
+There appears to be no requirement for the files indexed by the registry to be
+served from the same host as the registry, so the actual package files will be
+hosted externally.
+
+#### Package Hosting
+
+The distribution files (i.e. the `zip`, checksum and signature files) indexed
+by the registry will be stored in an AWS S3 bucket, and exposed to the Internet
+via a CloudFront CDN service.
+
+The storage bucket will be configured as per the Gold-class storage requirements
+described in the [Cloud Artifact Storage Standards](https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md)
+RFD.
+
+#### Why can't we just use Houston?
+
+The existing Houston distribution tool at [get.gravitational.com](https://get.gravitational.com/)
+is not suitable for distributing the files indexed by the registry. Houston has
+_very_ particular requirements about the naming of the files it serves (e.g.
+all files are expected to be `*.tar.gz` compressed tarballs - or at least
+_named_ as such) and, as it stands, will not serve the files we need it to.
+
+We _could_ modify Houston to serve the required files, but as the service is
+already slated for retirement, that does not seem to be a good use of time
+& money.
+
+### Building provider packages
+
+As far as the registry protocol is concerned, the entire package for a given
+provider consists of:
+
+* The actual binary distribution zipfile
+* A checksum file containing the SHA256 of the zipfile (formatted as per
+ `sha256sum` tool)
+* A signature file containing a binary, detached GPG signature for
+  checksum file.
+
+The Teleport provider release process currently produces a compressed tarball
+by default, and this should still be considered the primary artefact produced
+by the build, at least as long as Houston is expected to serve downloads of
+the Terraform provider.
+
+By analogy with the main Teleport release system, the registry-compatible
+provider package will be created during release promotion, with a build task
+that will:
+
+1. For each release tarball (one each for multiple release platforms):
+    1. repack the tarball into an appropriately formatted & named zipfile
+    2. generate the checksum file from the zipfile
+    3. generate the signature file by signing the checksum file
+2. Upload all generated files to the AWS bucket that backs the registry
+3. Update the registry index to include the newly released providers (see
+   _Populating the registry_, below)
+
+I expect this to be implemented as a small Go program, rather than a shell
+script. Using small Go programs for build tasks has shown itself to be a
+useful, flexible and above all _legible_ tool for scripting build tasks in
+the mainline Teleport repository, and we should continue that pattern in
+`teleport-plugins`.
+
+#### Build Security issues
```

Comment:
Can you explicitly call out permissions the pipeline will need on the buckets. For example, it should be able to write assets but not delete/modify anything. Also, enabling some sort of monitoring on S3 buckets (CloudWatch) may be useful too.

Reply:
https://github.com/gravitational/cloud/blob/master/rfd/0017-artifact-storage-standards.md#gold-1 should cover most of this.  For examples, you can grab some policy/terraform from https://github.com/gravitational/cloud/blob/master/rfd/assets/0017-artifact-storage-standards/artifact-bucket.tf